### PR TITLE
Enhance testbed-cli.sh script to support caching of deployed topologies

### PR DESCRIPTION
**Pull request by liorghub**
### Description of PR

Summary:
This PR replaces PR https://github.com/sonic-net/sonic-mgmt/pull/5674

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012

#### What is the motivation for this PR?
Enhance testbed-cli.sh script to support caching deployed topologies.
This is to reduce time when a developer wants to deploy a certain topology on a testbed, so only cached topology will be removed instead of all known topologies.

#### How did you do it?
Add a caching mechanism, each time a user ask to add a topology using the testbed-cli script, it will be cached.
Once the user will ask to deploy a different topology using testbed-cli script, only the cached topology will be removed and the new topology will be configured.
The cached topologies will be saved on a shared location known to all sonic-mgmt dockers and servers.
The path of the shared location should be provided as a separate file named "cached_topologies_path" under the ansible directory.
This file should be provided by each vendor on the internal git.
This PR provide the infrastructure to have this functionality, if the file contains the path is missing, the script will abort.

#### How did you verify/test it?
Replace topologies on a testbed.

### CODE REVIEW

**Comment by liorghub** [comment]
This PR replaces PR https://github.com/sonic-net/sonic-mgmt/pull/5674

**Comment by liorghub** [comment]
Failure in removing interface ip address operation.
 "stderr": "Cannot find device \"Ethernet112\""
Trying to rerun.

```
2022-08-17T06:52:22.9162326Z Traceback (most recent call last):
2022-08-17T06:52:22.9163667Z   File "/usr/local/bin/pytest", line 11, in <module>
2022-08-17T06:52:22.9165073Z     load_entry_point('pytest==4.6.5', 'console_scripts', 'pytest')()
2022-08-17T06:52:22.9166549Z   File "/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py", line 82, in main
2022-08-17T06:52:22.9167579Z     return config.hook.pytest_cmdline_main(config=config)
2022-08-17T06:52:22.9168975Z   File "/usr/local/lib/python2.7/dist-packages/pluggy/hooks.py", line 286, in __call__
2022-08-17T06:52:22.9170008Z     return self._hookexec(self, self.get_hookimpls(), kwargs)
2022-08-17T06:52:22.9171450Z   File "/usr/local/lib/python2.7/dist-packages/pluggy/manager.py", line 93, in _hookexec
2022-08-17T06:52:22.9173328Z     return self._inner_hookexec(hook, methods, kwargs)
2022-08-17T06:52:22.9174925Z   File "/usr/local/lib/python2.7/dist-packages/pluggy/manager.py", line 87, in <lambda>
2022-08-17T06:52:22.9175969Z     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
2022-08-17T06:52:22.9177540Z   File "/usr/local/lib/python2.7/dist-packages/pluggy/callers.py", line 208, in _multicall
2022-08-17T06:52:22.9178454Z     return outcome.get_result()
2022-08-17T06:52:22.9179732Z   File "/usr/local/lib/python2.7/dist-packages/pluggy/callers.py", line 81, in get_result
2022-08-17T06:52:22.9180624Z     _reraise(*ex)  # noqa
2022-08-17T06:52:22.9181890Z   File "/usr/local/lib/python2.7/dist-packages/pluggy/callers.py", line 187, in _multicall
2022-08-17T06:52:22.9183132Z     res = hook_impl.function(*args)
2022-08-17T06:52:22.9184652Z   File "/usr/local/lib/python2.7/dist-packages/_pytest/main.py", line 243, in pytest_cmdline_main
2022-08-17T06:52:22.9185643Z     return wrap_session(config, _main)
2022-08-17T06:52:22.9186980Z   File "/usr/local/lib/python2.7/dist-packages/_pytest/main.py", line 236, in wrap_session
2022-08-17T06:52:22.9187967Z     session=session, exitstatus=session.exitstatus
2022-08-17T06:52:22.9189336Z   File "/usr/local/lib/python2.7/dist-packages/pluggy/hooks.py", line 286, in __call__
2022-08-17T06:52:22.9190370Z     return self._hookexec(self, self.get_hookimpls(), kwargs)
2022-08-17T06:52:22.9191794Z   File "/usr/local/lib/python2.7/dist-packages/pluggy/manager.py", line 93, in _hookexec
2022-08-17T06:52:22.9193432Z     return self._inner_hookexec(hook, methods, kwargs)
2022-08-17T06:52:22.9194959Z   File "/usr/local/lib/python2.7/dist-packages/pluggy/manager.py", line 87, in <lambda>
2022-08-17T06:52:22.9196050Z     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
2022-08-17T06:52:22.9197540Z   File "/usr/local/lib/python2.7/dist-packages/pluggy/callers.py", line 203, in _multicall
2022-08-17T06:52:22.9198886Z     gen.send(outcome)
2022-08-17T06:52:22.9200576Z   File "/usr/local/lib/python2.7/dist-packages/_pytest/terminal.py", line 662, in pytest_sessionfinish
2022-08-17T06:52:22.9218640Z     outcome.get_result()
2022-08-17T06:52:22.9220131Z   File "/usr/local/lib/python2.7/dist-packages/pluggy/callers.py", line 81, in get_result
2022-08-17T06:52:22.9229748Z     _reraise(*ex)  # noqa
2022-08-17T06:52:22.9231401Z   File "/usr/local/lib/python2.7/dist-packages/pluggy/callers.py", line 187, in _multicall
2022-08-17T06:52:22.9232332Z     res = hook_impl.function(*args)
2022-08-17T06:52:22.9241492Z   File "/usr/local/lib/python2.7/dist-packages/_pytest/runner.py", line 73, in pytest_sessionfinish
2022-08-17T06:52:22.9242450Z     session._setupstate.teardown_all()
2022-08-17T06:52:22.9244043Z   File "/usr/local/lib/python2.7/dist-packages/_pytest/runner.py", line 325, in teardown_all
2022-08-17T06:52:22.9270673Z     self._pop_and_teardown()
2022-08-17T06:52:22.9275692Z   File "/usr/local/lib/python2.7/dist-packages/_pytest/runner.py", line 297, in _pop_and_teardown
2022-08-17T06:52:22.9276361Z     self._teardown_with_finalization(colitem)
2022-08-17T06:52:22.9277217Z   File "/usr/local/lib/python2.7/dist-packages/_pytest/runner.py", line 315, in _teardown_with_finalization
2022-08-17T06:52:22.9277764Z     self._callfinalizers(colitem)
2022-08-17T06:52:22.9279086Z   File "/usr/local/lib/python2.7/dist-packages/_pytest/runner.py", line 312, in _callfinalizers
2022-08-17T06:52:22.9279556Z     six.reraise(*exc)
2022-08-17T06:52:22.9280264Z   File "/usr/local/lib/python2.7/dist-packages/_pytest/runner.py", line 305, in _callfinalizers
2022-08-17T06:52:22.9281493Z     fin()
2022-08-17T06:52:22.9282259Z   File "/usr/local/lib/python2.7/dist-packages/_pytest/fixtures.py", line 859, in finish
2022-08-17T06:52:22.9339989Z     six.reraise(*e)
2022-08-17T06:52:22.9342951Z   File "/usr/local/lib/python2.7/dist-packages/_pytest/fixtures.py", line 852, in finish
2022-08-17T06:52:22.9347247Z     func()
2022-08-17T06:52:22.9352042Z   File "/usr/local/lib/python2.7/dist-packages/_pytest/fixtures.py", line 803, in _teardown_yield_fixture
2022-08-17T06:52:22.9352963Z     next(it)
2022-08-17T06:52:22.9354746Z   File "/var/src/s/tests/bgp/conftest.py", line 442, in setup_interfaces
2022-08-17T06:52:22.9355639Z     yield connections
2022-08-17T06:52:22.9357093Z   File "/usr/lib/python2.7/contextlib.py", line 24, in __exit__
2022-08-17T06:52:22.9374994Z     self.gen.next()
2022-08-17T06:52:22.9375846Z   File "/var/src/s/tests/bgp/conftest.py", line 422, in _setup_interfaces_t1_or_t2
2022-08-17T06:52:22.9377479Z     duthost.shell("config interface %s ip remove %s %s" % (namespace, conn["local_intf"], conn["local_addr"]))
2022-08-17T06:52:22.9378711Z   File "/var/src/s/tests/common/devices/multi_asic.py", line 100, in _run_on_asics
2022-08-17T06:52:22.9380393Z     return getattr(self.sonichost, self.multi_asic_attr)(*module_args, **complex_args)
2022-08-17T06:52:22.9381446Z   File "/var/src/s/tests/common/devices/base.py", line 89, in _run
2022-08-17T06:52:22.9411484Z     raise RunAnsibleModuleFail("run module {} failed".format(self.module_name), res)
2022-08-17T06:52:22.9432311Z tests.common.errors.RunAnsibleModuleFailLoading callback plugin json of type stdout, v2.0 from /usr/local/lib/python2.7/dist-packages/ansible/plugins/callback/json.pyc
2022-08-17T06:52:22.9442267Z : run module shell failed, Ansible Results =>
2022-08-17T06:52:22.9443126Z {
2022-08-17T06:52:22.9443819Z     "changed": true, 
2022-08-17T06:52:22.9444572Z     "cmd": "config interface  ip remove Ethernet112 20.0.0.0/31", 
2022-08-17T06:52:22.9445008Z     "delta": "0:00:02.627291", 
2022-08-17T06:52:22.9445796Z     "end": "2022-08-17 06:52:16.426603", 
2022-08-17T06:52:22.9446171Z     "failed": true, 
2022-08-17T06:52:22.9446475Z     "invocation": {
2022-08-17T06:52:22.9446779Z         "module_args": {
2022-08-17T06:52:22.9447208Z             "_raw_params": "config interface  ip remove Ethernet112 20.0.0.0/31", 
2022-08-17T06:52:22.9447647Z             "_uses_shell": true, 
2022-08-17T06:52:22.9447949Z             "argv": null, 
2022-08-17T06:52:22.9448253Z             "chdir": null, 
2022-08-17T06:52:22.9448564Z             "creates": null, 
2022-08-17T06:52:22.9448884Z             "executable": null, 
2022-08-17T06:52:22.9449189Z             "removes": null, 
2022-08-17T06:52:22.9449511Z             "stdin": null, 
2022-08-17T06:52:22.9449844Z             "stdin_add_newline": true, 
2022-08-17T06:52:22.9450184Z             "strip_empty_ends": true, 
2022-08-17T06:52:22.9450507Z             "warn": true
2022-08-17T06:52:22.9450780Z         }
2022-08-17T06:52:22.9451014Z     }, 
2022-08-17T06:52:22.9451522Z     "msg": "non-zero return code", 
2022-08-17T06:52:22.9451846Z     "rc": 1, 
2022-08-17T06:52:22.9452381Z     "start": "2022-08-17 06:52:13.799312", 
2022-08-17T06:52:22.9452803Z     "stderr": "Cannot find device \"Ethernet112\"", 
2022-08-17T06:52:22.9453173Z     "stderr_lines": [
2022-08-17T06:52:22.9453522Z         "Cannot find device \"Ethernet112\""
2022-08-17T06:52:22.9453845Z     ], 
2022-08-17T06:52:22.9454702Z     "stdout": "", 
2022-08-17T06:52:22.9455004Z     "stdout_lines": []
2022-08-17T06:52:22.9455261Z }
2022-08-17T06:52:23.5081515Z === Cleaning up DUT after tests ===
```

**Comment by liat-grozovik** [comment]
@wangxin please note that this PR was done prev by Shlomi but he is no longer avaialble to complete the work. So, new PR was introduced taking care your prev feedback.
Appreciate if you can review this one isntead.

**Comment by liat-grozovik** [comment]
@wangxin kindly reminder to review and signoff

File: ansible/testbed-cli.sh [hunk 1]
```diff
@@ -644,6 +652,103 @@ function cleanup_vmhost
...
+      for topo in $setup_topologies
+      do
+          ./testbed-cli.sh remove-topo $topo $passwd
+      done
+  else
+      .testbed-cli.sh remove-topo $cache_topo $passwd
+  fi
+
+  ./testbed-cli.sh add-topo $testbed_name $passwd
+  ./testbed-cli.sh deploy-mg $testbed_name $inventory $passwd
```
Comment (wangxin) [inline ansible/testbed-cli.sh:747]: Why call the script itself here? Why not call function `add_topo` and `deploy_minigraph ` instead?

Comment (wangxin) [reply ansible/testbed-cli.sh:747]: @liorghub Can you elaborate this?

Comment (liorghub) [reply ansible/testbed-cli.sh]: No need to call the script itself, I have changed it to function calls.

**Review by wangxin — APPROVED** [review]

**Review by roysr-nv — APPROVED** [review]

**Comment by liorghub** [comment]
Test tacacs/test_rw_user.py::test_rw_user failed. 
Trying to rerun.
```
2022-09-21T13:35:02.0839120Z common/devices/multi_asic.py:365: Failed
2022-09-21T13:35:02.0839793Z - generated xml file: /var/src/s/tests/logs/multi_asic_t1_lag/tacacs/test_rw_user.xml -
2022-09-21T13:35:02.0840252Z =========================== short test summary info ============================
2022-09-21T13:35:02.0840898Z ERROR tacacs/test_rw_user.py::test_rw_user[vlab-08] - Failed: ASIC instance n...
2022-09-21T13:35:02.0841351Z ========================== 1 error in 430.93 seconds ===========================
2022-09-21T13:35:02.1968686Z ##[error]Bash exited with code '1'.
2022-09-21T13:35:02.2009792Z ##[section]Finishing: Run tests
2022-09-21T13:35:02.2045320Z ##[section]Starting: Collect kvmdump
```

**Pull request merged**